system/security/LdapSecurity/ldapconnection.cpp

```diff
@@ -1121,30 +1121,40 @@ class CLdapClient : public CInterface, implements ILdapClient
     virtual void init(IPermissionProcessor* pp)
     {
         m_pp = pp;
-        if(m_ldapconfig->getServerType() == OPEN_LDAP)
+        static CriticalSection  lcCrit;
+        static bool createdOU = false;
+        if (!createdOU)
         {
-            try
-            {
-                addDC(m_ldapconfig->getBasedn());
-            }
-            catch(...)
-            {
-            }
-            try
-            {
-                addGroup("Directory Administrators", m_ldapconfig->getBasedn());
-            }
-            catch(...)
+            CriticalBlock block(lcCrit);
+            if (!createdOU)
```

Comment:
The potential problem is that the createdOU inside the critical block does not need to be re-evaluated - the compiler can use the value that was tested outside the critical block.

The cleanest fix is to use std::atomic_bool as the declaration for createdOU, and then it is guaranteed to be reloaded.  However how time critical is this function?  If it isn't it should be sufficient to always check createdOU inside the critical block.